File: fixtures/dotnet/j12/Fixture.cs
// Synthetic e-commerce codebase for explanation task
// Condensed representation of ~40 files

// === Program.cs (Main Entry Point) ===
// var builder = WebApplication.CreateBuilder(args);
// builder.Services.AddControllers();
// builder.Services.AddDbContext<AppDbContext>();
// builder.Services.AddStackExchangeRedisCache(...);
// var app = builder.Build();
// app.MapControllers();
// Routes registered via [ApiController] attributes:
//   /api/auth -> AuthController
//   /api/products -> ProductController
//   /api/cart -> CartController (requires CartMiddleware)
//   /api/checkout -> CheckoutController (requires AuthMiddleware + CartMiddleware)

// === Middleware/AuthMiddleware.cs ===
// Verifies JWT token from Authorization header
// Sets HttpContext.Items["User"] = { Id, Email, Role } if valid
// Returns 401 if token missing or expired
// public class AuthMiddleware {
//     public async Task InvokeAsync(HttpContext context, IConfiguration config) {
//         var token = context.Request.Headers["Authorization"].ToString();
//         if (string.IsNullOrEmpty(token) || !token.StartsWith("Bearer ")) {
//             context.Response.StatusCode = 401; return;
//         }
//         try {
//             var handler = new JwtSecurityTokenHandler();
//             var principal = handler.ValidateToken(token.Substring(7), validationParams, out _);
//             context.Items["User"] = principal.Claims;
//             await _next(context);
//         } catch { context.Response.StatusCode = 401; }
//     }
// }

// === Controllers/AuthController.cs ===
// POST /api/auth/register -- creates user, returns JWT
// POST /api/auth/login -- validates credentials, returns JWT
// POST /api/auth/refresh -- accepts refresh token, returns new access token
// Passwords hashed with BCrypt.Net (workFactor=12)
// JWT expiry: 1h access, 7d refresh

// === Controllers/ProductController.cs ===
// GET /api/products -- list with pagination (?page=1&pageSize=20)
// GET /api/products/{id} -- single product
// POST /api/products -- admin only ([Authorize(Roles = "Admin")])
// PUT /api/products/{id} -- admin only
// DELETE /api/products/{id} -- admin only (soft delete, sets DeletedAt)
// No caching implemented -- direct DB queries on every request (TECH DEBT)

// === Controllers/CartController.cs ===
// Cart stored in Redis with key cart:{userId}
// GET /api/cart -- returns current cart items
// POST /api/cart/items -- add item (checks inventory via ProductRepository)
// DELETE /api/cart/items/{productId} -- remove item
// Inventory not reserved during cart session -- race condition possible (TECH DEBT)

// === Controllers/CheckoutController.cs ===
// POST /api/checkout -- processes cart into order
// Validates inventory at checkout time
// Creates Order + OrderItems in a database transaction (using IDbContextTransaction)
// Calls external payment processor (Stripe via PaymentService)
// Sends confirmation email (no retry on failure -- TECH DEBT)

// === Models/User.cs ===
// public class User { ... }
// Id (int), Email (unique), PasswordHash, Role (User|Admin), CreatedAt, LastLogin
// EF Core entity, PostgreSQL backend (Npgsql)

// === Models/Product.cs ===
// public class Product { ... }
// Id (int), Name, Description, Price (decimal), Stock (int), CategoryId, DeletedAt (soft delete)
// No database indexes on CategoryId or Price (TECH DEBT)

// === Models/Order.cs ===
// public class Order { ... }
// Id (int), UserId, Status (Pending|Paid|Shipped|Delivered|Cancelled), Total (decimal), CreatedAt
// ICollection<OrderItem> Items navigation property

// === Models/OrderItem.cs ===
// public class OrderItem { ... }
// Id (int), OrderId, ProductId, Quantity, UnitPrice (decimal, snapshot at purchase time)

// === Services/PaymentService.cs ===
// Wraps Stripe.net SDK
// CreateCharge(amount, currency, source) -- returns ChargeResult { ChargeId, Status }
// No idempotency keys implemented (TECH DEBT -- duplicate charges possible)

// === Data/AppDbContext.cs ===
// PostgreSQL via Npgsql (EF Core provider)
// Connection pool: MinPoolSize=2, MaxPoolSize=10 (hardcoded in connection string, not configurable via appsettings)

// === Configuration/RedisConfig.cs ===
// Redis client for cart storage via StackExchange.Redis
// No connection error handling -- app crashes if Redis unavailable

class Fixture
{
    // placeholder for the condensed codebase representation
}

File: fixtures/nodejs/j12/fixture.js
// Synthetic e-commerce codebase for explanation task
// Condensed representation of ~40 files

// === server.js ===
// const express = require('express');
// const authRoutes = require('./routes/auth');
// const productRoutes = require('./routes/products');
// const cartRoutes = require('./routes/cart');
// const checkoutRoutes = require('./routes/checkout');
// app.use('/api/auth', authRoutes);
// app.use('/api/products', productRoutes);
// app.use('/api/cart', cartMiddleware, cartRoutes);
// app.use('/api/checkout', authMiddleware, cartMiddleware, checkoutRoutes);

// === middleware/auth.js ===
// Verifies JWT token from Authorization header
// Sets req.user = { id, email, role } if valid
// Returns 401 if token missing or expired
// const jwt = require('jsonwebtoken');
// function authMiddleware(req, res, next) {
//   const token = req.headers.authorization?.split(' ')[1];
//   if (!token) return res.status(401).json({ error: 'No token' });
//   try {
//     req.user = jwt.verify(token, process.env.JWT_SECRET);
//     next();
//   } catch { res.status(401).json({ error: 'Invalid token' }); }
// }

// === routes/auth.js ===
// POST /api/auth/register — creates user, returns JWT
// POST /api/auth/login — validates credentials, returns JWT
// POST /api/auth/refresh — accepts refresh token, returns new access token
// Passwords hashed with bcrypt (rounds=12)
// JWT expiry: 1h access, 7d refresh

// === routes/products.js ===
// GET /api/products — list with pagination (?page=1&limit=20)
// GET /api/products/:id — single product
// POST /api/products — admin only (role check in route)
// PATCH /api/products/:id — admin only
// DELETE /api/products/:id — admin only (soft delete, sets deleted_at)
// No caching implemented — direct DB queries on every request (TECH DEBT)

// === routes/cart.js ===
// Cart stored in Redis with key cart:{userId}
// GET /api/cart — returns current cart items
// POST /api/cart/items — add item (checks inventory)
// DELETE /api/cart/items/:productId — remove item
// Inventory not reserved during cart session — race condition possible (TECH DEBT)

// === routes/checkout.js ===
// POST /api/checkout — processes cart into order
// Validates inventory at checkout time
// Creates Order + OrderItems in a DB transaction
// Calls external payment processor (Stripe)
// Sends confirmation email (no retry on failure — TECH DEBT)

// === models/User.js ===
// id, email (unique), passwordHash, role (user|admin), createdAt, lastLogin
// Sequelize model, PostgreSQL backend

// === models/Product.js ===
// id, name, description, price, stock, categoryId, deletedAt (soft delete)
// No database indexes on categoryId or price (TECH DEBT)

// === models/Order.js ===
// id, userId, status (pending|paid|shipped|delivered|cancelled), total, createdAt
// hasMany OrderItems

// === models/OrderItem.js ===
// id, orderId, productId, quantity, unitPrice (snapshot at purchase time)

// === services/PaymentService.js ===
// Wraps Stripe API
// createCharge(amount, currency, source) — returns { chargeId, status }
// No idempotency keys implemented (TECH DEBT — duplicate charges possible)

// === config/database.js ===
// PostgreSQL via Sequelize
// Connection pool: min=2, max=10 (hardcoded, not configurable via env)

// === config/redis.js ===
// Redis client for cart storage
// No connection error handling — app crashes if Redis unavailable

module.exports = {}; // placeholder for the condensed codebase representation
